distribution/lib/Standard/Base/0.0.0-dev/src/Data/XML.enso

```diff
@@ -0,0 +1,389 @@
+    attribute self name:Text ~if_missing=(Error.throw (No_Such_Key.Error self name)) =
+        XML_Error.handle_java_exceptions <|
+            attr = self.java_element.getAttributeNode name
+            if attr.is_nothing then if_missing else attr.getValue
+
+    ## Gets a map containing f the attributes of an XML element.
+
+       > Example
+           XML_Document.from_string '<foo bar="one">hello</foo>' . root_element . attributes
+           # => Map.from_vector [["bar", "one"]]
+    attributes : Map Text Text ! XML_Error
+    attributes self =
+        XML_Error.handle_java_exceptions <|
+            named_node_map = self.java_element.getAttributes
+            keys_and_values = 0.up_to named_node_map.getLength . map i->
+                node = named_node_map.item i
+                [node.getNodeName, node.getNodeValue]
+            Map.from_vector keys_and_values
+
+    ## Gets the text (non-markup) contents of the element and its descendants,
+       if any.
+
+       > Example
+         Get the text content of an element.
+
+             XML_Document.from_string '<foo>hello</foo>' . root_element . text
+             # => "hello"
+    text : Text ! XML_Error
+    text self =
+        XML_Error.handle_java_exceptions <|
+            self.java_element.getTextContent
+
+    ## Gets the raw XML of the element (including tag, attributes and contents).
+
+       > Example
+         Get the outer XML of an element.
+
+             XML_Document.from_string '<foo>hello</foo>' . root_element . outer_xml
+             # => '<foo>hello</foo>'
+    outer_xml : Text ! XML_Error
+    outer_xml self =
+        XML_Error.handle_java_exceptions <|
+            XML_Utils.outerXML self.java_element
+
+    ## Gets the raw XML of the contents of the element, not including the
+       outermost tag and attributes.
+
+       > Example
+         Get the inner XML of an element.
+
+             XML_Document.from_string '<foo><bar>hello</bar></foo>' . root_element . inner_xml
+             # => '<bar>hello</bar>'
+    inner_xml : Text ! XML_Error
+    inner_xml self =
+        XML_Error.handle_java_exceptions <|
+            XML_Utils.innerXML self.java_element
+
+    ## Gets elements matching a given tag name.
+
+       This searches through all descendants of the node, not just immediate children.
+
+       > Example
+             XML_Document.from_string '<foo> <baz>hello</baz> <bar>and</bar> <baz>goodbye</baz> </foo>' . root_element . get_elements_by_tag_name "baz"
+             # => [XML_Document.from_string "<baz>hello</baz>" . root_element, XML_Document.from_string "<baz>goodbye</baz>" . root_element]
+    get_elements_by_tag_name : Text -> Vector XML_Element ! XML_Error
+    get_elements_by_tag_name self tag_name:Text =
+        XML_Error.handle_java_exceptions <|
+            only_wanted_nodes (nodelist_to_vector (self.java_element.getElementsByTagName tag_name))
+
+    ## PRIVATE
+    Value (java_element:Element)
```

Comment:
As with XML_Document we need to make sure this renders in the table.

Reply:
Done